infra/modules/codepipeline/codepipeline.tf

```diff
@@ -0,0 +1,190 @@
+data "aws_ssm_parameter" "github_codestar_connection_arn" {
+  name = "${var.project_name}_github_codestar_connection_arn"
+}
+
+resource "aws_codepipeline" "docker-image-codepipeline" {
+  name          = "${var.project_name}-docker-image-pipeline-${var.environment}"
+  role_arn      = aws_iam_role.codepipeline_role.arn
+  pipeline_type = "V2"
+
+  artifact_store {
+    type     = "S3"
+    location = aws_s3_bucket.codepipeline_bucket.bucket
+  }
+
+  stage {
+    name = "Source"
+
+    action {
+      name             = "Source"
+      category         = "Source"
+      owner            = "AWS"
+      provider         = "CodeStarSourceConnection"
+      version          = "1"
+      output_artifacts = ["source_output"]
+      namespace        = "SourceVariables"
+
+      // options given here: https://docs.aws.amazon.com/codepipeline/latest/userguide/action-reference-CodestarConnectionSource.html#action-reference-CodestarConnectionSource-config
+      configuration = {
+        ConnectionArn        = data.aws_ssm_parameter.github_codestar_connection_arn.value
+        FullRepositoryId     = "JiscSD/octopus"
+        BranchName           = var.environment
+        OutputArtifactFormat = "CODE_ZIP"
+      }
+    }
+  }
+
+  stage {
+    name = "Deploy-Image"
+
+    action {
+      name            = "Deploy-Image"
+      category        = "Build"
+      owner           = "AWS"
+      provider        = "CodeBuild"
+      input_artifacts = ["source_output"]
+      version         = "1"
+
+      configuration = {
+        ProjectName = aws_codebuild_project.deploy-docker-image.name
+        EnvironmentVariables = jsonencode([
+          {
+            name  = "COMMIT_ID",
+            type  = "PLAINTEXT"
+            value = "#{SourceVariables.CommitId}"
+          }
+        ])
+      }
+    }
+  }
+}
+
+resource "aws_iam_role" "codepipeline_role" {
+  name = "${var.project_name}-codepipeline-role-${var.environment}"
+  tags = {
+    Name = "${var.project_name}-codepipeline-role-${var.environment}"
+  }
+
+  assume_role_policy = <<EOF
```

Comment:
I'm probably guilty as anyone of inconsistent defining of policies using EOF or jsonencode. This suggests another way which means the policy is checked fro syntax errors before you apply

https://cloudonaut.io/defining-iam-policies-with-terraform/

in any case, be consistent...

Reply:
also here https://github.com/hashicorp/terraform-provider-aws/issues/17714#issuecomment-1418274657

so seems better to go with the separate definition style..

somewhere I need to start a terraform style guide..